Finer topic handling for SUB/SYNC_SUB

Flood1993 opened the pull request:
I still need to squash lots of commits and refactor a bit (maybe some parameter names and method signatures).

Flood1993 commented:
As of now, internally, all handlers for SUB sockets are stored as a dictionary, of the form `{topic: handler, ...}`

For the special SUB case, we have to set the socket to listen to that topic, and also store the handler internally so we know what must be called when a message with that topic is received. In bb7e5a8, a bug occurs in which by overriding the handlers, we lose the first topic.

A real life example of that would be osmarkets, in which we might connect to different pairs at any point in time.

See #159

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +73,164 @@ def on_error(agent):
+    # Message not received
+    server.send('pub', 'not received')
+    assert not wait_agent_attr(client, data='not received',
+                               timeout=0.5)
```
There might be a sync problem in this test I think: maybe we can send from the publisher before the client completed the unsubscription?

Maybe we could use the incremental numbers trick? Publish two topics `a`/`b`, then unsubscribe from one of them. Then wait for the last received value difference (`a` - `b`) to be greater than 1?

Maybe we can create a `wait_for` function, which accepts a function to execute periodically on the agent until it succeeds or the timeout is reached.

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +73,164 @@ def on_error(agent):
+    # Only message from subscribed topic received
+    server.send('pub', 1, topic='negate')
+    server.send('pub', 2, topic='normal')
+    server.send('pub', 'not received', topic='topic_C')
```
Could we use `.each()` to avoid sync issues in case subscription is not completed?

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +73,164 @@ def on_error(agent):
+    assert not wait_agent_attr(client, data=1, timeout=0.5)
+    assert not wait_agent_attr(client, data=-2, timeout=0.5)
+    assert not wait_agent_attr(client, data='not received',
+                               timeout=0.5)
```
I think the first two `assert not wait` might be redundant. But if you wanted to keep them, maybe set the timeout to `0.5` only for the first `assert not wait`. Then for the rest set it to `0` to make it a bit faster?

Peque commented on osbrain/agent.py:
```diff
@@ -930,11 +933,59 @@ def _subscribe(self, alias: str, handlers: Dict[Union[bytes, str], Any]):
+            curated_handlers = topics_to_bytes(handlers, uuid=uuid)
+            self._set_handler(self.socket[sub_address], curated_handlers)
+        else:
+            self._set_handler(self.socket[alias], curated_handlers)
```
Maybe this is why it fails. It seems we are settings `curated_handlers`, but what you actually want to do is to update this dictionary if it existed before? (not sure, did not run the tests)

Flood1993 commented:
I'm wondering if it would be a good idea to make `_set_handler` a non private method. That way, the user could easily modify the handlers for other communication patterns other than PUB/SUB and SYNC_PUB/SYNC_SUB. What do you think @Peque ?

Flood1993 commented:
I have improved the tests a bit so as to improve their consistency and readability.

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +74,152 @@ def on_error(agent):
+    client.connect(addr, alias='sub', handler={'odd': append_received,
+                                               'even': append_received})
+
+    # Make sure the connection is stablished
```
"established"?

Flood1993 replied:
😅

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +74,152 @@ def on_error(agent):
+
+
+@pytest.mark.parametrize('socket_type', ['PUB', 'SYNC_PUB'])
+def test_resubscribe(nsproxy, socket_type):
```
Not sure if this `test_resubscribe` is an interesting/required test. Even if so, I would call it `test_unsubscribe_and_subscribe_again`.

For me, a `test_resubscribe` would be one that actually checked the behavior in this case:

```python
client.subscribe('sub', handler={'x': foo})
client.subscribe('sub', handler={'x': bar})
```

Subscription after subscription, without an `unsubscribe` call in between.

This is currently not defined behavior. In my opinion we could simply override the first subscription. What you think?

Flood1993 replied:
I agree with everything you said. I renamed the test, and I will work now on writing the true ``test_resubscribe``. I will comment again when I'm done with it. 

For me, I would leave the ``test_unsubscribe_and_subscribe_again`` in the test code base, as it won't hurt to have more tests. What do you think?

Flood1993 replied:
Implemented the new test - see commit 9d44e10.

Peque replied:
Yeah, we can leave it there. :blush:

Peque commented on osbrain/agent.py:
```diff
@@ -933,9 +944,11 @@ def _handle_async_requests(self, data):
 
-    def _subscribe(self, alias: str, handlers: Dict[Union[bytes, str], Any]):
+    def subscribe(self, alias: str,
+                  handlers: Dict[Union[bytes, str], Any]) -> None:
```
I think we should use `handler` (singular) to be consistent with `bind`/`connect`. What you think?

Flood1993 replied:
I agree. Changed all occurrences!

Flood1993 commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +74,152 @@ def on_error(agent):
+    # Wait until client receives some numbers
+    client.set_attr(received=[])
+    server.each(0.1, publish)
+    assert wait_agent_attr(client, length=5)
```
Yes, maybe it would be easier to read and understand the test. Maybe we can change it towards the end of the PR? To get the most important things tackled first.

Flood1993 commented:
I pushed various commits with the following changes:
- Remove specifying the topics as bytes. You are right, it makes more sense to use strings plus the use of bytes is undocumented.
- I changed the tests to implement the new functionality you proposed, checking the tail of the messages.
- From the last point, I refactored the server agents to share the same timer for their publications, so as to try to guarantee that the tails are always ``...1, 2]``, or at least for most of the time, because it gave me problems keeping the timers separated.
- I also changed some strings from the test. I feel many of them are still meaningless, I feel like I should try to make the tests more readable for future maintenance, but I'm not sure if the proper way is with in-line comments or in the docstring of the test.

Let me know what you think @Peque so we can keep working on this PR 😄

Flood1993 commented:
Uhh, I don't know why it's crashing on Travis (at least I'm able to reproduce the crashes in my Ubuntu VM). In Windows all tests are passing OK. I will investigate and come back when I find something.

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -59,6 +60,13 @@ def reply(self, delay):
+        self.each(0.1, 'send', 'pub', 1, topic='negate')
+        self.each(0.1, 'send', 'pub', 2, topic='normal')
+        self.each(0.1, 'send', 'pub', 'not received', topic='topic_C')
+
```
If you want to make sure messages are sent in a given order, why don't you do:

```python
def publish(self):
    self.send('pub', 1, topic='negate')
    self.send('pub', 2, topic='normal')
    self.send('pub', 'not received', topic='topic_C')

# [...]
each(0.1, 'publish')
```

?

Just started from here my review, so maybe I'll find later something that makes me understand this better... :sweat_smile:

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +81,175 @@ def on_error(agent):
 
 
+def match_tail(agent, length, data):
+    return agent.received[-length:] == data
```
Or even...

```python
def match_tail(agent, data):
￼     return agent.received[-len(data):] == data
```

?

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +81,175 @@ def on_error(agent):
+    client.unsubscribe('sub', 'odd')
+
+    assert wait_agent_condition(client, check_non_consecutive)
+    assert client.get_attr('received')[-1] % 2 == 0
```
Looks good. I would, however, check that `assert not client.execute_as_method(check_non_consecutive)` before unsubscribing. Just to make sure we were receiving consecutive numbers before and that we are not now.

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +81,175 @@ def on_error(agent):
+    client.set_attr(received=[])
+
+    # Nothing should be received (length == 0)
+    assert not wait_agent_attr(client, length=1, timeout=0.5)
```
Why don't you unsubscribe from one of the topics only so that you can do:

```python
assert wait_agent_attr(client, match_tail, [-1] * 8)
```

instead?

That means, positive assertion, with no need for a timeout (less risk). You can decrease the publisher period to `0.01` to make things faster.

Also, this way you do not need to reset the `received` attribute. :wink:

Flood1993 replied:
At first I did what you proposed, but then the code should be ``client.unsubscribe('sub', 'normal')`` 😅 . So I changed it for the other way around, just to avoid any subliminal messages hahaha.

Peque commented:
@Flood1993 I just suggested you exactly that (it looked suspicious). :joy:

Peque commented on docs/source/basic_patterns.rst:
```diff
@@ -176,3 +176,29 @@ subscribed to topic ``a`` only and *Dave* subscribed to topic ``b`` only.
+   the handlers on the ``connect()``/``bind()`` call of the SUB socket.
+
+.. note:: Calls to the ``subscribe()`` method will always override the
+   previous handler for each of the specified topics, if any.
```
Maybe this could be a warning?

Peque commented on docs/source/basic_patterns.rst:
```diff
@@ -176,3 +176,29 @@ subscribed to topic ``a`` only and *Dave* subscribed to topic ``b`` only.
+communication.
+
+It is also remarkable that we can easily modify the subscriptions at runtime
+easily.
```
Maybe we can simplify this two paragraphs and say it like: "We can also very easily modify the subscriptions at run-time." (says the same with less words?).

Peque commented on osbrain/agent.py:
```diff
@@ -933,29 +944,82 @@ def _handle_async_requests(self, data):
+        if isinstance(topic, (tuple, list, dict)):
+            for t in topic:
+                self.unsubscribe(alias, t)
+            return
```
This is not currently tested. Could you add a test? And also maybe update the docstring to say that a list/tuple is accepted in order to unsubscribe from multiple topics.

Peque commented on osbrain/tests/test_agent_sync_publications.py:
```diff
@@ -73,6 +81,173 @@ def on_error(agent):
+        Check whether the last two received numbers are non-consecutive.
+        """
+        if len(agent.received) < 2:
+            return False
```
This (`if len(agent.received) < 2`) seems to never happen. If it does not, then remove it. Otherwise make sure it always happens to ensure maximum coverage of our tests as well. :wink:

Peque commented:
@Flood1993 Pushed one commit here to avoid setting the channel's UUID with a non-unique identifier (it kind of violates the whole purpose of a unique identifier :joy:).

Two tests fail due to the "shutdown many agents test". That was expected after the last shutdown modification. See #208.

Apart from that there are only a couple of minor comments from yesterday and this should be ready for merge. :blush:

Peque commented:
@Flood1993 A fix for the "shutdown many agents test" is read in #209. Seems to work fine.

Flood1993 commented:
@Peque I did the changes to the documentation and also added a test for multiple unsubscriptions in the same call. I removed the option to pass the topics as a dictionary since it seems counter intuitive to me. Do you think I should leave it there?

Peque commented:
@Flood1993 Seems `test_agent_sync_publications.py` coverage was reduced. You can correct those lambdas. Try with:

```python
addr = server.bind(..., lambda agent, message: None)
```

Or maybe even better, use an existing handler:

```python
addr = server.bind(..., handler=append_receive)
```

Peque commented:
@Flood1993 There are also two lines not branch-covered in `agent.py`:

![screenshot from 2017-11-10 12-34-44](https://user-images.githubusercontent.com/526577/32656822-ace7bf66-c613-11e7-8d60-4beea724b23d.png)

Could you have a look at that? I would prefer to merge only code that either keeps or improves code coverage, but never decreases it (unless very special circumstances).
